waspc/waspls/src/Wasp/LSP/SignatureHelp.hs

```diff
@@ -0,0 +1,261 @@
+          logM "[getSignatureHelpAtPosition] no signature found"
+          return emptyHelp
+        Just signature -> do
+          logM "[getSignatureHelpAtPosition] found a signature"
+          let signatureFragments = signatureToFragments signature
+          let params = getParamsFromFragments signatureFragments
+          let activeParam = signatureParam signature >>= findActiveParameterIndex signatureFragments
+          logM $ "[getSignatureHelpAtPosition] at param idx = " ++ show activeParam ++ " params = " ++ show params
+          return $
+            LSP.SignatureHelp
+              { _signatures =
+                  LSP.List
+                    [ LSP.SignatureInformation
+                        { _parameters = Just $ LSP.List params,
+                          -- NOTE: VSCode higlights the 0th parameter if
+                          -- activeParameter is Nothing, so it's set to -1 when
+                          -- we don't want any param highlighted.
+                          _activeParameter = Just $ fromMaybe (-1) activeParam,
+                          _label = Text.pack $ showFragments signatureFragments,
+                          _documentation = Nothing
+                        }
+                    ],
+                _activeSignature = Just 0,
+                _activeParameter = Nothing
+              }
+  where
+    emptyHelp =
+      LSP.SignatureHelp
+        { _signatures = LSP.List [],
+          _activeSignature = Nothing,
+          _activeParameter = Nothing
+        }
+
+-- | 'Signature' describes the expected type at a specific location inside of
+-- a container in a wasp file. Every signature includes the type of the container
+-- surrounding the location. For example, the container type for a cursor
+-- positioned inside of a dictionary would be the type of the dictionary.
+--
+-- When the location is at a more specific place inside of the container--for
+-- example, immediately after @key: @ in a dictionary--then we say the location
+-- is at a parameter, and the parameter is described in relation to the container
+-- with an 'ExprKey'.
+data Signature
+  = -- | Inside a container, but not at any parameter inside of it.
+    InContainer !Type
+  | -- | Inside a container at the parameter specified by the 'ExprKey'. This
+    -- is named \"param\" to be consistent with the naming conventions used by
+    -- the LSP.
+    AtParam !Type !ExprKey
+  deriving (Eq, Show)
+
+signatureType :: Signature -> Type
+signatureType (InContainer typ) = typ
+signatureType (AtParam typ _) = typ
+
+signatureParam :: Signature -> Maybe ExprKey
+signatureParam (InContainer _) = Nothing
+signatureParam (AtParam _ key) = Just key
+
+-- | @'findSignatureAtLocation' sourceCode location@ runs type inference at the
+-- given location to try to find a 'Signature'.
+--
+-- To do this, two types are inferred: the /tip type/ and the /parent type/. The
+-- tip type is the type expected exactly at the cursor location and the parent
```

Comment:
Why tipType, why not locationType or smth like that?

Reply:
How would "tip" fit into the tree methapor though?

Reply:
Ok, sounds good! Btw isn't tip of the branch usually called "leaf"? Or did you have something different in mind?

Reply:
Yeah that's true. I can't recall any reason I chose tip instead of leaf.